File: 2022/june/21-06-2022/furthest-building-you-can-reach.py
class Solution:
    def furthestBuilding(self, heights, bricks: int, ladders: int) -> int:
        heights_count = len(heights)

        def abc(current_position, curr_bricks, curr_ladders):
            if current_position > heights_count:
                return current_position

            if bricks <= curr_bricks and ladders <= curr_ladders:
                return current_position
            
            if heights[current_position] > heights[current_position+1]:
                return abc(current_position+1, curr_bricks, curr_ladders)

            if bricks > curr_bricks:
                with_bricks = abc(
                    current_position+1, 
                    curr_bricks + (heights[current_position+1]-heights[current_position]), 
                    curr_ladders
                )
            else:
                with_bricks = 0

            if ladders > curr_ladders:
                with_ladders = abc(
                    current_position+1, 
                    curr_bricks, 
                    curr_ladders+1
                )
            else:
                with_ladders = 0

            return max(with_bricks, with_ladders)

        return abc(0, 0, 0)


print(Solution().furthestBuilding([4,2,7,6,9,14,12], 5, 1))

File: 2022/june/21-06-2022/furthest-building-you-can-reach.rs
use std::collections::BinaryHeap;

impl Solution {
    pub fn furthest_building(heights: Vec<i32>, mut bricks: i32, mut ladders: i32) -> i32 {
        let mut last_bldg = i32::MAX;
        let mut count = -1;
        let mut pq = BinaryHeap::new();
        for bldg in heights {
            let diff = bldg - last_bldg;
            if diff < 0 {
                last_bldg = bldg;
                count += 1;
                continue
            }
            if diff <= bricks {
                pq.push(diff);
                bricks -= diff;
                count += 1;
                last_bldg = bldg;
                continue
            }
            if ladders > 0 {
                if !pq.is_empty() {
                    let reclaimed_bricks = pq.pop().unwrap(); // already checked that okay to unwrap b/c not empty
                    if reclaimed_bricks > diff {
                        pq.push(diff);
                        bricks = bricks + reclaimed_bricks - diff;
                    } else {
                        pq.push(reclaimed_bricks);
                        // I don't know if more efficient to pop() then push() if not useable or to peek() then pop() if useable
                    }
                }
                ladders -= 1;
                count += 1;
                last_bldg = bldg;
                continue
            }
            return count
        }
        count
    }
}
